### PR TITLE
[SYCL] Fix memory leak for sycl::device::get_device

### PR DESCRIPTION
**Pull request by bso-intel**
Plugins create pi_device objects for root devices, but
it does not call piDeviceRelease and causes memory leaks.
This slows the performance of SYCL apps.

Signed-off-by: Byoungro So <byoungro.so@intel.com>

### CODE REVIEW

File: sycl/source/detail/device_impl.cpp [hunk 1]
```diff
@@ -74,7 +74,7 @@ device_impl::device_impl(pi_native_handle InteropDeviceHandle,
 }
 
 device_impl::~device_impl() {
-  if (!MIsRootDevice && !MIsHostDevice) {
```
Comment (vladimirlaz) [inline sycl/source/detail/device_impl.cpp:77]: Looks like the failure was caused by your change:
/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.obj/tools/sycl/unittests/scheduler/./SchedulerTests(+0xb6d5a)[0x556a1871fd5a]
/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.obj/tools/sycl/unittests/scheduler/./SchedulerTests(+0xb4f64)[0x556a1871df64]
/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.obj/tools/sycl/unittests/scheduler/./SchedulerTests(+0xb50b3)[0x556a1871e0b3]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x12890)[0x7f8d30ed7890]
/rdrive/ref/opencl/runtime/linux/oclgpu/20.21.16886/libze_intel_gpu.so.0.8(_ZN2L014CommandQueueHwIL14GFXCORE_FAMILY12EE7destroyEv+0x25)[0x7f8d2ef43375]
/rdrive/ref/opencl/runtime/linux/oclgpu/20.21.16886/libze_intel_gpu.so.0.8(_ZN2L011CommandListD1Ev+0x30)[0x7f8d2ef49890]
/rdrive/ref/opencl/runtime/linux/oclgpu/20.21.16886/libze_intel_gpu.so.0.8(_ZN2L033CommandListImmediateProductFamilyIL14PRODUCT_FAMILY20EED0Ev+0x17)[0x7f8d2ef3ee77]
/rdrive/ref/opencl/runtime/linux/oclgpu/20.21.16886/libze_intel_gpu.so.0.8(_ZN2L014CommandListImp7destroyEv+0x13)[0x7f8d2ef49c13]
**/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.obj/./lib/libpi_level0.so(piDeviceRelease+0x39)[0x7f8d2f741cc9]**
/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.obj/tools/sycl/unittests/scheduler/./SchedulerTests(+0x20f75)[0x556a18689f75]
/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.obj/tools/sycl/unittests/scheduler/./SchedulerTests(+0x1ff6a)[0x556a18688f6a]
/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.obj/tools/sycl/unittests/scheduler/./SchedulerTests(+0x9baea)[0x556a18704aea]
/lib/x86_64-linux-gnu/libc.so.6(+0x43041)[0x7f8d2fba8041]
/lib/x86_64-linux-gnu/libc.so.6(+0x4313a)[0x7f8d2fba813a]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xee)[0x7f8d2fb86b9e]
/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.obj/tools/sycl/unittests/scheduler/./SchedulerTests(+0x15c3a)[0x556a1867ec3a]

Comment (bso-intel) [reply sycl/source/detail/device_impl.cpp:77]: This is expected failure due to the issue in L0 RT.
The L0 RT issue is now fixed, and we are still waiting for their release (soon).
I will rerun this failure once the new L0 RT is released.

File: buildbot/dependency.conf [hunk 2]
```diff
@@ -5,8 +5,8 @@ ocl_cpu_rt_ver=2020.10.6.0.04
 ocl_cpu_rt_ver_win=2020.10.6.0.04
 # Same GPU driver supports Level Zero and OpenCL:
 # https://github.com/intel/compute-runtime/releases/tag/20.21.16886
-ocl_gpu_rt_ver=20.21.16886
-# Same GPU driver supports Level Zero and OpenCL:
+ocl_gpu_rt_ver=20.26.17199
+# Same GPU driver spports Level Zero and OpenCL:
```
Comment (bader) [inline buildbot/dependency.conf:9]:
```suggestion
# Same GPU driver supports Level Zero and OpenCL:
```

**Comment by bso-intel** [comment]
> sycl/* part is OK, could you please add a test?

Thanks for reviewing. A test is added.

**Comment by bso-intel** [comment]
> Please, fix regressions reported by CI system.

CI system will continue to fail until the QA team upgrades the GPU RT to the latest version in the CI system. Please hold merging this PR until this work is complete.

File: buildbot/dependency.conf [hunk 3]
```diff
@@ -5,7 +5,7 @@ ocl_cpu_rt_ver=2020.10.6.0.04
 ocl_cpu_rt_ver_win=2020.10.6.0.04
 # Same GPU driver supports Level Zero and OpenCL:
 # https://github.com/intel/compute-runtime/releases/tag/20.21.16886
-ocl_gpu_rt_ver=20.21.16886
+ocl_gpu_rt_ver=20.26.17199
```
Comment (vladimirlaz) [inline buildbot/dependency.conf:8]: please update link above when you change driver version.

File: sycl/test/basic_tests/memory-consumption.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: env SYCL_BE=PI_LEVEL0 %t.out
```
Comment (vladimirlaz) [inline sycl/test/basic_tests/memory-consumption.cpp:2]: That will not work on machines without OpenCL or Level0 RTs.
please update RUN lines to make them functionol in all possible configurations

File: buildbot/dependency.conf (hunk 3 above)
Comment (bso-intel) [reply buildbot/dependency.conf:8]: done

File: sycl/test/basic_tests/memory-consumption.cpp (hunk 4 above)
Comment (bso-intel) [reply sycl/test/basic_tests/memory-consumption.cpp]: done

File: sycl/test/basic_tests/memory-consumption.cpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  memset(&usage, 0, sizeof(rusage));
+  getrusage(RUSAGE_SELF, &usage);
+  return usage.ru_maxrss;
+}
+#endif
+
+using namespace cl::sycl;
+
+int main() {
+  constexpr auto dev_type = info::device_type::gpu;
```
Comment (vladimirlaz) [inline sycl/test/basic_tests/memory-consumption.cpp:37]: Your test is GPU-oriented isn't it better to check all possible devices?
If we want to test all available devices we can use default device_selector and RUN lines like in https://github.com/intel/llvm/blob/sycl/sycl/test/sub_group/info.cpp#L3-L7

File: sycl/test/basic_tests/memory-consumption.cpp [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  int startSize = get_cpu_mem();
+  std::cout << startSize << " kb" << std::endl;
+
+  for (int i = 0; i < 1000; i++) {
+    devices = cl::sycl::device::get_devices(dev_type);
+  }
+  int endSize = get_cpu_mem();
+  std::cout << endSize << " kb" << std::endl;
+#else
+  for (int i = 0; i < 1000; i++) {
```
Comment (vladimirlaz) [inline sycl/test/basic_tests/memory-consumption.cpp:50]: We do not check memory for windows is there any reason to run the test on windows?
I would prefer to disable it there by putting something like
// No way to count memory
// UNSUPPORTED: windows

File: sycl/test/basic_tests/memory-consumption.cpp (hunk 5 above)
Comment (bso-intel) [reply sycl/test/basic_tests/memory-consumption.cpp:38]: > Your test is GPU-oriented isn't it better to check all possible devices?
> If we want to test all available devices we can use default device_selector and RUN lines like in https://github.com/intel/llvm/blob/sycl/sycl/test/sub_group/info.cpp#L3-L7

This test is meant to test the memory leak caused by not calling plugin APIs.
It does not make sense to test all possible devices.

File: sycl/test/basic_tests/memory-consumption.cpp (hunk 6 above)
Comment (bso-intel) [reply sycl/test/basic_tests/memory-consumption.cpp:51]: Done

Comment (vladimirlaz) [reply sycl/test/basic_tests/memory-consumption.cpp:51]: please remove dead code as well

File: sycl/test/basic_tests/memory-consumption.cpp [hunk 7]
```diff
@@ -0,0 +1,70 @@
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: env SYCL_BE=%sycl_be %t.out
```
Comment (vladimirlaz) [inline sycl/test/basic_tests/memory-consumption.cpp:3]: can be removed
```suggestion
```

**Comment by bso-intel** [comment]
I wonder why rebasing from upstream includes all previous history as a changeset in PR.
It makes very difficult for the reviewers to see which are my own changes.
I will close this PR and recreate a new PR for review.

**Comment by bader** [comment]
You need just re-create a new history in your `dt-devices-leak` branch and you can continue using this PR.
Please, use `git merge` instead of `git rebase` to sync with the `sycl` branch.

**Pull request closed without merging**